File: testing/web/tests/engine/spacing/MarginExtractor.test.ts
// Tests for MarginExtractor — covers every IR shape from the margin fixtures.
import { describe, it, expect } from 'vitest';
import { extractMargin } from '../../../src/style/engine/spacing/MarginExtractor';
import { applyMargin } from '../../../src/style/engine/spacing/MarginApplier';

// Helper to wrap data into IRProperty-ish record.
const p = (type: string, data: unknown) => ({ type, data });

describe('extractMargin', () => {
  it('handles absolute px margin', () => {
    const cfg = extractMargin([p('MarginTop', { px: 20 })]);
    expect(cfg.top).toEqual({ kind: 'exact', px: 20 });
  });

  it('handles negative px margin (shifts layout)', () => {
    const cfg = extractMargin([p('MarginLeft', { px: -10 })]);
    expect(cfg.left).toEqual({ kind: 'exact', px: -10 });
  });

  it('handles bare "auto" string (flex centering)', () => {
    const cfg = extractMargin([p('MarginLeft', 'auto'), p('MarginRight', 'auto')]);
    expect(cfg.left).toEqual({ kind: 'auto' });
    expect(cfg.right).toEqual({ kind: 'auto' });
  });

  it('handles mixed auto + px (margin: auto 0 scenario)', () => {
    const cfg = extractMargin([
      p('MarginTop', 'auto'), p('MarginRight', { px: 0 }),
      p('MarginBottom', 'auto'), p('MarginLeft', { px: 0 }),
    ]);
    expect(cfg.top).toEqual({ kind: 'auto' });
    expect(cfg.right).toEqual({ kind: 'exact', px: 0 });
  });

  it('handles bare number as percentage (Margin_Percent_10 fixture)', () => {
    const cfg = extractMargin([p('MarginTop', 10)]);
    expect(cfg.top).toEqual({ kind: 'relative', value: 10, unit: 'percent' });
  });

  it('handles em margin (runtime-dependent)', () => {
    const cfg = extractMargin([p('MarginRight', { original: { v: 2, u: 'EM' } })]);
    expect(cfg.right).toEqual({ kind: 'relative', value: 2, unit: 'em' });
  });

  it('handles all logical margin sides', () => {
    const cfg = extractMargin([
      p('MarginBlockStart',  { px: 1 }),
      p('MarginInlineEnd',   'auto'),
    ]);
    expect(cfg.blockStart?.kind).toBe('exact');
    expect(cfg.inlineEnd).toEqual({ kind: 'auto' });
  });

  it('ignores non-margin IR properties', () => {
    const cfg = extractMargin([p('PaddingTop', { px: 10 })]);
    expect(cfg).toEqual({});
  });
});

describe('applyMargin', () => {
  it('emits auto keyword verbatim', () => {
    const css = applyMargin({ left: { kind: 'auto' }, right: { kind: 'auto' } });
    expect(css).toEqual({ marginLeft: 'auto', marginRight: 'auto' });
  });

  it('emits negative px', () => {
    const css = applyMargin({ top: { kind: 'exact', px: -10 } });
    expect(css).toEqual({ marginTop: '-10px' });
  });

  it('emits percent and logical sides', () => {
    const css = applyMargin({
      top: { kind: 'relative', value: 10, unit: 'percent' },
      blockEnd: { kind: 'exact', px: 4 },
    });
    expect(css.marginTop).toBe('10%');
    expect(css.marginBlockEnd).toBe('4px');
  });
});

File: runtimes/compose/src/test/java/com/styleconverter/runtime/spacing/MarginExtractorTest.kt
package com.styleconverter.runtime.spacing

// Unit tests for MarginExtractor. Emphasis on the auto/negative/mixed cases
// that PaddingExtractor doesn't need to worry about.

import com.styleconverter.runtime.core.types.LengthUnit
import com.styleconverter.runtime.core.types.LengthValue
import kotlinx.serialization.json.Json
import org.junit.Assert.assertEquals
import org.junit.Assert.assertNull
import org.junit.Assert.assertTrue
import org.junit.Test

class MarginExtractorTest {

    private fun parse(s: String) = Json.parseToJsonElement(s)
    private fun pair(type: String, json: String) = type to parse(json)

    @Test fun `absolute px on all four sides`() {
        val cfg = MarginExtractor.extract(listOf(
            pair("MarginTop", """{"px":20.0}"""),
            pair("MarginRight", """{"px":30.0}"""),
            pair("MarginBottom", """{"px":40.0}"""),
            pair("MarginLeft", """{"px":50.0}"""),
        ))
        assertEquals(MarginValue.Length(LengthValue.Exact(20.0)), cfg.top)
        assertEquals(MarginValue.Length(LengthValue.Exact(50.0)), cfg.left)
        assertTrue(cfg.hasMargin)
    }

    @Test fun `negative px preserved as Length with negative Exact`() {
        val cfg = MarginExtractor.extract(listOf(
            pair("MarginTop", """{"px":-10.0}"""),
        ))
        assertEquals(MarginValue.Length(LengthValue.Exact(-10.0)), cfg.top)
    }

    @Test fun `auto keyword maps to MarginValue-Auto`() {
        val cfg = MarginExtractor.extract(listOf(
            pair("MarginLeft", """"auto""""),
        ))
        assertEquals(MarginValue.Auto, cfg.left)
        assertTrue(cfg.hasHorizontalAuto)
    }

    @Test fun `margin auto 0 mixed auto and zero length`() {
        // CSS `margin: 0 auto` → top/bottom {px:0}, left/right "auto".
        val cfg = MarginExtractor.extract(listOf(
            pair("MarginTop", """{"px":0.0}"""),
            pair("MarginRight", """"auto""""),
            pair("MarginBottom", """{"px":0.0}"""),
            pair("MarginLeft", """"auto""""),
        ))
        assertEquals(MarginValue.Length(LengthValue.Exact(0.0)), cfg.top)
        assertEquals(MarginValue.Auto, cfg.left)
        assertEquals(MarginValue.Auto, cfg.right)
        assertTrue(cfg.hasHorizontalAuto)
        assertEquals(false, cfg.hasVerticalAuto)
    }

    @Test fun `em relative unit propagates through MarginValue-Length`() {
        val cfg = MarginExtractor.extract(listOf(
            pair("MarginTop", """{"original":{"v":1.5,"u":"EM"}}"""),
        ))
        assertEquals(
            MarginValue.Length(LengthValue.Relative(1.5, LengthUnit.EM, null)),
            cfg.top,
        )
    }

    @Test fun `bare number margin treated as percent`() {
        val cfg = MarginExtractor.extract(listOf(
            pair("MarginTop", "10.0"),
        ))
        assertEquals(
            MarginValue.Length(LengthValue.Relative(10.0, LengthUnit.PERCENT, null)),
            cfg.top,
        )
    }

    @Test fun `logical MarginBlockStart stored separately from Top`() {
        val cfg = MarginExtractor.extract(listOf(
            pair("MarginBlockStart", """{"px":4.0}"""),
        ))
        assertEquals(MarginValue.Length(LengthValue.Exact(4.0)), cfg.blockStart)
        assertNull(cfg.top)
    }

    // ── wave-45 lane X4: the prebaked-em preservation ────────────────────
    // DynamicValueResolver's pre-pass adds a resolved px NEXT TO an em
    // original (shape `{"px":N,"original":{"v":V,"u":"EM"}}`); the
    // extractor must keep the em identity (with the prebake as fallback)
    // so the applier can re-resolve against the element's OWN computed
    // font-size (css-values-4 §5.1.1 — the CSS Fonts 4 §3.5 monospace-13
    // quirk case measured on wave-44 discard-multicol-001).

    @Test fun `prebaked em keeps Relative with the prebake as pxFallback`() {
        // The exact post-resolver wire of discard-multicol-001's div
        // (margin: 1em prebaked at the 16px default basis).
        val cfg = MarginExtractor.extract(listOf(
            pair("MarginTop", """{"px":16.0,"original":{"v":1.0,"u":"EM"}}"""),
        ))
        assertEquals(
            MarginValue.Length(LengthValue.Relative(1.0, LengthUnit.EM, 16.0)),
            cfg.top,
        )
    }

    @Test fun `prebaked em inside the typed length wrapper is preserved too`() {
        // The `{"type":"length"}` envelope variant the pre-pass preserves.
        val cfg = MarginExtractor.extract(listOf(
            pair("MarginTop", """{"type":"length","px":46.0,"original":{"v":0.5,"u":"EM"}}"""),
        ))
        assertEquals(
            MarginValue.Length(LengthValue.Relative(0.5, LengthUnit.EM, 46.0)),
            cfg.top,
        )
    }

    @Test fun `prebaked NON-em originals keep the canonical Exact px`() {
        // Absolute source units (converter emits px beside e.g. PT): px is
        // canonical, exactly as before wave 45 — only EM re-resolves.
        val cfg = MarginExtractor.extract(listOf(
            pair("MarginTop", """{"px":16.0,"original":{"v":12.0,"u":"PT"}}"""),
            pair("MarginBottom", """{"px":39.0,"original":{"v":10.0,"u":"VW"}}"""),
        ))
        assertEquals(MarginValue.Length(LengthValue.Exact(16.0)), cfg.top)
        assertEquals(MarginValue.Length(LengthValue.Exact(39.0)), cfg.bottom)
    }

    @Test fun `isMarginProperty matches all 8 longhands`() {
        val all = listOf(
            "MarginTop", "MarginRight", "MarginBottom", "MarginLeft",
            "MarginBlockStart", "MarginBlockEnd",
            "MarginInlineStart", "MarginInlineEnd",
        )
        for (name in all) assertTrue(name, MarginExtractor.isMarginProperty(name))
        assertEquals(false, MarginExtractor.isMarginProperty("PaddingTop"))
    }
}
